src/server/browserType.ts

```diff
@@ -86,42 +88,88 @@ export abstract class BrowserTypeBase implements BrowserType {
 
   async launch(options: LaunchOptions = {}): Promise<Browser> {
     assert(!(options as any).userDataDir, 'userDataDir option is not supported in `browserType.launch`. Use `browserType.launchPersistentContext` instead');
-    const browserServer = new BrowserServer(options);
-    const { transport, downloadsPath } = await this._launchServer(options, 'local', browserServer);
-    return await browserServer._initializeOrClose(async () => {
-      return this._connectToServer(browserServer, false, transport!, downloadsPath);
-    });
+    return this._innerLaunch('local', options);
   }
 
   async launchPersistentContext(userDataDir: string, options: LaunchOptions = {}): Promise<BrowserContext> {
-    const browserServer = new BrowserServer(options);
-    const { transport, downloadsPath } = await this._launchServer(options, 'persistent', browserServer, userDataDir);
+    const browser = await this._innerLaunch('persistent', options, userDataDir);
+    return browser._defaultContext!;
+  }
+
+  async _innerLaunch(launchType: LaunchType, options: LaunchOptions, userDataDir?: string): Promise<BrowserBase> {
+    const deadline = TimeoutSettings.computeDeadline(options.timeout, 30000);
+    const logger = new RootLogger(options.logger);
+    logger.startLaunchRecording();
+
+    let browserServer: BrowserServer | undefined;
+    try {
+      browserServer = await this._launchServer(options, launchType, logger, deadline, userDataDir);
+      const promise = this._innerLaunchPromise(browserServer, launchType, options);
+      const browser = await helper.waitWithDeadline(promise, 'the browser to launch', deadline, 'pw:browser*');
+      return browser;
+    } catch (e) {
+      e.message += '\n=============== Process output during launch: ===============\n' +
```

Comment:
Should we do this after _closeOrKill to capture it's output as well?

Reply:
I don't think so - it adds one more failure point.